rfcs/text/0000-alternative-planning.md

```diff
@@ -0,0 +1,315 @@
+  
+Nowadays it's pretty much orthodoxy that nobody does waterfall unless they must (a contentious statement perhaps
+though I've seen that sentiment in RFC reviews). So why is it being proposed by the Planning and Roadmaps RFC? From my 
+perspective it's because of these challenges:
+
+* **High complexity** - the DCP feels highly complex. This makes for a very difficult management problem. Everything is
+connected.
+* **Siloed teams** - A team works on and owns its own components. Getting other teams to act means negotiation. It's
+ much easier to do a local solution that is sub-optimal for the DCP as a whole than a global one where some teams may 
+ need to change their chosen design.
+* **Constrained communication** - plenary voice communication is tightly constrained to an hour per day. This constrains
+communication both between DCP groups with the same role (e.g. ingest, analysis, azul dev groups) and up and down
+the objective to implementation work stack (e.g. between developers and product owners, between product owners and 
+project leads).
+* **No overall control** - no organization is in overall control.
+* **Cultural variation** - organization's planning and development culture varies.
+
+The response in the Roadmaps and Planning RFC, for me, is a waterfall one of long negotiation and tight alignment. 
+As a result, if one strictly follows the process then design and implementation become separated by long stretches of time.
+Software suffers from the problems of waterfall described above.
+
+When reality meets an ill-fitting process something has to give. As a practical consequence, I think that agile will take place
+"under the covers". It will happen because it has to in order to get things done but despite the process rather than
+because of it. We already see this today where in reality, implementation of 
+RFCs has to start before they are approved in order to meet deadlines in an unpredictable environment and gain 
+knowledge about whether the approach will really work. With work such as reliability, we simply don't know what makes
+sense until we do other work first - it's just not possible to plan a long time in advance if you don't want to waste
+time confabulating detailed written plans that you know aren't worth the pixels they're displayed on.
+
+I am definitely not saying that planning and RFCs are not worthwhile. There is no question that we face complex problems
+that need advance thinking. But I think that the generally feeling is that we're heading towards a place where there are simultaneously too many RFCs for
+anybody to review and not enough to tackle the complex problems that we face.
+
+### Alternatives
+
+So what could be the alternatives? I am going to propose several actions which I think could be implemented separately but
+ may be most powerful if they were all to be implemented together.
+
+#### Management by aims and indicators
```

Comment:
I've thought about this possibility for some time now after @tburdett first suggested it. I think there is a fundamental challenge that DCP has that other organizations who practice agile do not: _our microservices are pipelined a deeply nested architecture_.

Compare two software architectures and their affects on coordination:

1. At a prior company we had a "flat" architecture. The important backend services didn't regularly depend on each other and fed a back-end-for-front-end (BEFE) gateway: `ServiceA -> BEFE <- ServiceB...`. This meant that the node depth of the graph was 1 and you only ever had to change one edge in the graph to deliver new value to user. Teams could run agile scrums 90% of the time and when an edge needed to change they had a couple weeks of coordination with a predetermined migration plan.
2. Compare the prior example to the DCP where the path is `Ingest -> Upload -> DSS -> Analyisis -> Ingest -> Upload -> DSS -> {Browser, Matrix}. The node depth here is 7. Unless the architecture changes radically, this means that there will necessarily be more coordination required to deliver value to users unless you are `{Browser, Matrix}` and already have the data you need to deliver a feature.

[A reference](https://medium.learningbyshipping.com/plans-3bcc6164e8b) to back this up:
> If you’re coordinating plans across two orgs, then timing is the only universal language — and it is a language of commitment. Two teams can only work together if they share high integrity timing — two functional teams or client/back end, etc.

Reply:
@mweiden I think that is a very valuable insight. However, I would draw the conclusion that having such a great node depth means that we have fundamentally the wrong architecture. No amount of process, waterfall or otherwise, will fix that. It needs some overall architecture rethinking which is near impossible because of our per-component split.

Reply:
@justincc there's a bit of a catch-22 here. Options are 1) refactor the architecture (if it can be refactored) or 2) work with the current architecture. In either case, we need careful planning and coordination to move forward. Only once we've demonstrated that we have the organization and architecture to support independent development can we safely loosen our planning cycles.